ruby/ql/lib/codeql/ruby/frameworks/actioncontroller/Filters.qll

```diff
@@ -0,0 +1,360 @@
+    private FilterKind kind;
+
+    Filter() {
+      this = call.getFilterArgument() and
+      kind = call.getKind() and
+      call.getMethodName() = ["", "prepend_", "append_"] + kind + "_action"
+    }
+
+    predicate isPrepended() { call.getMethodName().regexpMatch("^prepend.+$") }
+
+    MethodCallCfgNode getCall() { result = call }
+
+    FilterKind getKind() { result = kind }
+
+    private ModuleBase getEnclosingModule() { result = call.getExpr().getEnclosingModule() }
+
+    /**
+     * Holds if this callback is registered before `other`. This does not
+     * guarantee that the callback will be executed before `other`. For example,
+     * `after_action` callbacks are executed in reverse order.
+     */
+    predicate registeredBefore(Filter other) {
+      // before_action :this, :other
+      //
+      // before_action :this
+      // before_action :other
+      this.getBasicBlock() = other.getBasicBlock() and this.getASuccessor+() = other
+      or
+      this.getEnclosingModule() = other.getEnclosingModule() and
+      this.getBasicBlock().strictlyDominates(other.getBasicBlock()) and
+      this != other
+      or
+      // This callback is in a superclass of `other`'s class.
+      //
+      // class A
+      //   before_action :this
+      //
+      // class B < A
+      //   before_action :other
+      other.getEnclosingModule().getModule() = this.getEnclosingModule().getModule().getASubClass+()
+    }
+
+    /**
+     * Holds if this callback runs before `other`.
+     */
+    private predicate runsBefore(Filter other, ActionControllerActionMethod action) {
+      other.getKind() = this.getKind() and
+      not this.skipped(action) and
+      not other.skipped(action) and
+      action = this.getAction() and
+      action = other.getAction() and
+      (
+        not this.isPrepended() and
+        (
+          not other.isPrepended() and
+          (
+            this.getKind() = TBeforeKind() and
+            this.registeredBefore(other)
+            or
+            this.getKind() = TAfterKind() and
+            other.registeredBefore(this)
+          )
+          or
+          other.isPrepended() and this.getKind() = TAfterKind()
+        )
+        or
+        this.isPrepended() and
+        (
+          not other.isPrepended() and
+          this.getKind() = TBeforeKind()
+          or
+          other.isPrepended() and
+          (
+            this.getKind() = TBeforeKind() and this.registeredBefore(other)
+            or
+            this.getKind() = TAfterKind() and other.registeredBefore(this)
+          )
+        )
+      )
+    }
+
+    Filter getNextFilter(ActionControllerActionMethod action) {
+      result != this and
+      this.runsBefore(result, action) and
+      not exists(Filter mid | this.runsBefore(mid, action) | mid.runsBefore(result, action))
+    }
+
+    predicate skipped(ActionControllerActionMethod action) {
+      this = any(SkipFilter f | f.getKind() = this.getKind()).getSkippedFilter(action)
```

Comment:
Does there need to be any regard for the ordering of the calls here? e.g. in the contrived case of:
```
  before_action :foo
  skip_before_action :foo
  before_action :foo
```
I would have naively assumed that `foo` would end up being registered as a before action.